File: context_tools/bdd/examples/examples.py
"""BDD examples — Mamba/Python. Index in examples.md; TypeScript in examples.ts."""

# Behavior fidelity — signatures (character_spec.py)

from mamba import description, context, it

with description('a Character'):
    with context('that has been created'):
        with it('should have initial stats assigned'):
            # BDD: SIGNATURE
            pass
        with it('should have zero starting wounds'):
            # BDD: SIGNATURE
            pass

    with context('that is in combat'):
        with context('that has suffered wounds'):
            with it('should have current wounds'):
                # BDD: SIGNATURE
                pass

        with context('that has been attacked with the following damage'):
            with it('should add the attack damage to current wounds'):
                # BDD: SIGNATURE
                pass

    with description('an Attack'):
        with context('that has targeted an enemy'):
            with it('should calculate hit chance using character stats'):
                # BDD: SIGNATURE
                pass
            with it('should consume one action from the active turn'):
                # BDD: SIGNATURE
                pass

        with context('that has missed'):
            with it('should deal no damage'):
                # BDD: SIGNATURE
                pass
            with it('should still consume one action'):
                # BDD: SIGNATURE
                pass

    with context('that has been defeated'):
        with it('should be removed from the initiative order'):
            # BDD: SIGNATURE
            pass


# Development fidelity — test implementation (character_spec.py)

from mamba import description, context, it, before
from expects import equal, expect
from character import Character


def default_stats():
    return {'strength': 10, 'agility': 8, 'endurance': 6}


with description('a Character'):
    with context('that has been created'):
        with it('should have initial stats assigned'):
            # Arrange / Act
            character = Character(name='Test', stats=default_stats())
            # Assert
            expect(character.stats['strength']).to(equal(10))
            expect(character.stats['agility']).to(equal(8))

        with it('should have zero starting wounds'):
            # Arrange / Act
            character = Character(name='Test', stats=default_stats())
            # Assert
            expect(character.wounds).to(equal(0))

    with context('that is in combat'):
        with before.each:
            self.character = Character(name='Test', stats=default_stats())

        with context('that has suffered wounds'):
            with it('should have current wounds'):
                self.character.apply_damage(3)
                expect(self.character.wounds).to(equal(3))

        with context('that has been attacked with the following damage'):
            with it('should add the attack damage to current wounds'):
                self.character.apply_damage(2)
                self.character.apply_damage(4)
                expect(self.character.wounds).to(equal(6))


# Development fidelity — minimal production code (character.py)


class Character:
    def __init__(self, name: str, stats: dict):
        self.name = name
        self.stats = stats
        self.wounds = 0

    def apply_damage(self, amount: int) -> None:
        self.wounds += amount

File: practices/bdd/examples/examples.ts
/**
 * BDD examples — Jest/TypeScript. Narrative and index live in examples.md.
 * Mamba/Python: examples.py
 */

// Behavior hierarchy input (prerequisite) — character-behavior.md
//
// a Character
//   that has been created
//     should have initial stats assigned
//     should have zero starting wounds
//   that is in combat
//     that has suffered wounds
//       should have current wounds
//     that has been attacked with the following damage
//       should add the attack damage to current wounds
//   an Attack
//     that has targeted an enemy
//       should calculate hit chance using character stats
//       should consume one action from the active turn
//     that has missed
//       should deal no damage
//       should still consume one action
//   that has been defeated
//     should be removed from the initiative order

// Behavior fidelity — signatures (character.test.ts)

describe('a Character', () => {
  describe('that has been created', () => {
    it('should have initial stats assigned', () => {
      // BDD: SIGNATURE
    });
    it('should have zero starting wounds', () => {
      // BDD: SIGNATURE
    });
  });

  describe('that is in combat', () => {
    describe('that has suffered wounds', () => {
      it('should have current wounds', () => {
        // BDD: SIGNATURE
      });
    });

    describe('that has been attacked with the following damage', () => {
      it('should add the attack damage to current wounds', () => {
        // BDD: SIGNATURE
      });
    });
  });

  describe('an Attack', () => {
    describe('that has targeted an enemy', () => {
      it('should calculate hit chance using character stats', () => {
        // BDD: SIGNATURE
      });
      it('should consume one action from the active turn', () => {
        // BDD: SIGNATURE
      });
    });

    describe('that has missed', () => {
      it('should deal no damage', () => {
        // BDD: SIGNATURE
      });
      it('should still consume one action', () => {
        // BDD: SIGNATURE
      });
    });
  });

  describe('that has been defeated', () => {
    it('should be removed from the initiative order', () => {
      // BDD: SIGNATURE
    });
  });
});

// Development fidelity — signature input (character.test.ts, partial)

describe('Character', () => {
  describe('that has been created', () => {
    it('should have initial stats assigned', () => {
      // BDD: SIGNATURE
    });
    it('should have zero starting wounds', () => {
      // BDD: SIGNATURE
    });
  });

  describe('that is in combat', () => {
    describe('that has suffered wounds', () => {
      it('should have current wounds', () => {
        // BDD: SIGNATURE
      });
    });

    describe('that has been attacked with the following damage', () => {
      it('should add the attack damage to current wounds', () => {
        // BDD: SIGNATURE
      });
    });
  });
});

// Development fidelity — test implementation (character.test.ts)

import { Character } from '../Character';

function defaultStats() {
  return { strength: 10, agility: 8, endurance: 6 };
}

describe('Character', () => {
  describe('that has been created', () => {
    it('should have initial stats assigned', () => {
      // Arrange
      const stats = defaultStats();
      // Act
      const character = new Character({ name: 'Test', stats });
      // Assert
      expect(character.stats.strength).toBe(10);
      expect(character.stats.agility).toBe(8);
    });

    it('should have zero starting wounds', () => {
      // Arrange / Act
      const character = new Character({ name: 'Test', stats: defaultStats() });
      // Assert
      expect(character.wounds).toBe(0);
    });
  });

  describe('that is in combat', () => {
    let character: Character;

    beforeEach(() => {
      character = new Character({ name: 'Test', stats: defaultStats() });
    });

    describe('that has suffered wounds', () => {
      it('should have current wounds', () => {
        // Act
        character.applyDamage(3);
        // Assert
        expect(character.wounds).toBe(3);
      });
    });

    describe('that has been attacked with the following damage', () => {
      it('should add the attack damage to current wounds', () => {
        // Arrange
        character.applyDamage(2);
        // Act
        character.applyDamage(4);
        // Assert
        expect(character.wounds).toBe(6);
      });
    });
  });
});

// Development fidelity — minimal production code (Character.ts)

interface Stats {
  strength: number;
  agility: number;
  endurance: number;
}

interface CharacterProps {
  name: string;
  stats: Stats;
}

export class Character {
  readonly name: string;
  readonly stats: Stats;
  wounds = 0;

  constructor({ name, stats }: CharacterProps) {
    this.name = name;
    this.stats = stats;
  }

  applyDamage(amount: number): void {
    this.wounds += amount;
  }
}

// Layer boundary mocking example (service layer)

import { VoucherService } from '../VoucherService';
import { VoucherRepository } from '../VoucherRepository';

describe('VoucherService', () => {
  describe('that is creating a voucher', () => {
    let service: VoucherService;
    let mockRepo: jest.Mocked<Pick<VoucherRepository, 'save'>>;

    beforeEach(() => {
      mockRepo = { save: jest.fn().mockResolvedValue(undefined) };
      service = new VoucherService(mockRepo as VoucherRepository);
    });

    it('should persist the voucher when input is valid', async () => {
      // Arrange
      const input = { code: 'ABC-001', campaignId: 'camp-1' };
      // Act
      await service.create(input);
      // Assert
      expect(mockRepo.save).toHaveBeenCalledWith(
        expect.objectContaining({ code: 'ABC-001' })
      );
    });
  });
});
